Bug 1178874: Add new sync_projects command.

Osmose opened the pull request:
Adds a new management command called `sync_projects` that simultaneously pulls updates from VCS to update the database and commits changes from the database to VCS.

More details to follow, and this is still slightly incomplete, but I figured putting it up now for some pre-reviewing would be helpful. Don't merge for now.

@mathjazz Comments, criticism, and questions are welcome. :D

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+
+
+class Command(BaseCommand):
+    args = '<project_id project_id ...>'
```
Shall we switch to using slug instead of id?

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+    def log(self, msg, *args, **kwargs):
+        """Log a message to the console."""
+        self.stdout.write(msg.format(*args, **kwargs) + u'\n')
+
```
Is that newline character at the end even necessary? https://docs.djangoproject.com/en/1.8/howto/custom-management-commands/

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+            self.log(msg, *args, **kwargs)
+
+    def debug(self, msg, *args, **kwargs):
+        """Log a message to the console if --verbosty=2."""
```
Typo: verbosty -> verbosity

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+
+        # Once we've synced, we can deleted all translations scheduled
+        # for deletion.
+        Translation.deleted_objects.all().delete()
```
Typo: we can deleted -> we can delete

mathjazz commented on pontoon/base/models.py:
```diff
@@ -201,6 +272,10 @@ class Entity(models.Model):
     @property
+    def unique_key(self):
+        return ':'.join([self.resource.path, self.string])
+
```
Some file formats (xliff, properties, dtd, inc, ini, l20n) use "key" as a unique string identifier in a resource file.

mathjazz commented on pontoon/base/vcs_models.py:
```diff
@@ -0,0 +1,107 @@
+    @property
+    def unique_key(self):
+        return ':'.join([self.resource.relative_path, self.string])
+
```
Some file formats (xliff, properties, dtd, inc, ini, l20n) use "key" as a unique string identifier in a resource file.

mathjazz commented on pontoon/base/models.py:
```diff
@@ -232,6 +307,24 @@ def serialize(self):
             'obsolete': self.obsolete,
         }
 
+    def has_changed(self, locale_code):
```
Would these querysets be evaluated only once instead of two times if you moved the .exists() piece right after .filter()? https://docs.djangoproject.com/en/1.8/ref/models/querysets/#when-querysets-are-evaluated

mathjazz replied:
locale_code is never used

Osmose replied:
> Would these querysets be evaluated only once instead of two times if you moved the .exists() piece right after .filter()?

They're not evaluated when `filter` is called because it's lazy, so they only get evaluated once when `exists` is called. Putting the call on the same line vs the line below is equivalent because assigning the queryset to a variable doesn't evaluate it.

mathjazz commented on pontoon/base/utils.py:
```diff
@@ -320,3 +320,16 @@ def req(method, project, resource, locale,
+        lambda i: all(getattr(i, attrib) == value
+                      for attrib, value in attributes.items()),
+        default=None
+    )
```
Adding short comments to both of these methods would be helpful.

mathjazz commented on pontoon/base/models.py:
```diff
@@ -1,15 +1,20 @@
 from django.core.exceptions import ValidationError
 from django.db import models
-from django.db.models import Sum
+from django.db.models import Q, Sum
```
Q is never used.

mathjazz commented on pontoon/base/formats/lang.py:
```diff
@@ -0,0 +1,133 @@
+        self.string = string
+        self.translations = {None: translation}
+        self.comments = comments
+        self.fuzzy = False
```
There is no concept of fuzzy in .lang file format.

Osmose replied:
Correct, but if we don't do this then the sync code that handles setting the fuzzy flag or reading it in the database would throw AttributeErrors when we try to access the property. So we fake it out on the entity and never use it when reading or writing to the lang files.

mathjazz commented on pontoon/base/vcs_models.py:
```diff
@@ -0,0 +1,107 @@
+                else:
+                    vcs_entity = VCSEntity(
+                        resource=self,
+                        string=format_entity.string,
```
Later on, we will also need plurals.

mathjazz commented on pontoon/base/vcs_models.py:
```diff
@@ -0,0 +1,107 @@
+class VCSEntity(object):
+    def __init__(self, resource, string, comments, order):
+        self.resource = resource
+        self.string = string
```
Later on, we will also need plurals.

Osmose replied:
Yeah, the plural support here is half-assed for now and will be fleshed out when we add a second format that supports proper plurals.

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+            resource, created = db_project.resource_set.get_or_create(path=relative_path)
+            resource.format = Resource.get_path_format(relative_path)
+            resource.entity_count = len(vcs_resource.entities)
+            resource.save()
```
Why aren't resource.project and resource.path also set?

Osmose replied:
Because we're using the manager from `db_project.resource_set`, the project is implicitly set to `db_project`, and the path is set by `get_or_create`.

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+                # This should never happen. What? Hard abort.
+                raise CommandError('No entities found for key {0}'.format(key))
+            else:
+                # VCS no longer has the entity, remove it from Pontoon.
```
We should set obsolete=True instead. Removing entity will also remove translations and thus affect translation memory and contribution stats.

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+        entities = (Entity.objects
+                    .select_related('resource')
+                    .filter(resource__project=db_project))
+        return {entity.unique_key: entity for entity in entities}
```
We should exclude obsolete=True entities.

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+
+        for locale in db_project.locales.all():
+            commit_message = (
+                'Pontoon: Update {locale.name}({locale.code}) localization of '
```
Add space before openning bracket.

mathjazz replied:
Also, we should probably say "Updated" instead of "Update".

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+            if result is not None:
+                self.log(
+                    'Committing project {project.name} for {locale.name}'
+                    '({locale.code}) failed: {reason}',
```
Add space before openning bracket.

mathjazz commented:
Very well done! I'd love to see more code comments (for methods) and more detailed description how this PR actually works. Some comments added inline, minor improvements only.

I've evaluated the code in my head against the following cases that have caused us problems recently:
- Strings that get unfuzzied get fuzzy again after update from VCS.
- If translation added to VCS, that is already stored as unapproved in Pontoon, it doesn't prevail.
- https://bugzilla.mozilla.org/show_bug.cgi?id=1178874#c2.

And it works just fine! Bravo!

A question on the new .lang file parser: since it's the only format parser that was part of Pontoon (stolen from bedrock), I suspect we won't be needing any such rewrites for other formats (which are handled by external libraries), except for the dump_* and extract_* replacemenets.

Also, shall we keep the line limit to max. 79 characters? I was trying to do that until now, but it's also causing headaches.

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,327 @@
+        for locale in db_project.locales.all():
+            commit_message = (
+                'Pontoon: Update {locale.name}({locale.code}) localization of '
+                '{project.name}.'
```
And no period at the end.

Osmose commented:
> A question on the new .lang file parser: since it's the only format parser that was part of Pontoon (stolen from bedrock), I suspect we won't be needing any such rewrites for other formats (which are handled by external libraries), except for the dump_* and extract_* replacemenets.

Correct, sort've. For the other formats we'll need to make sure they conform to the same API that we use to alter the lang files (where we have "format entities" that have attributes we can change, and a resource class that we can call `save()` on), but we should be able to have light wrappers around the existing parsers for those. Before I merge this I'd like to have that API specifically written out as some classes for the langfile classes to inherit from.

Osmose commented:
Updated. I've amended the previous commit with a proper message and made changes in response to your feedback.

Notably I've added a bunch of comments explaining some things, and added some classes to define the interface between `sync_projects`, `pontoon.base.formats`, and `pontoon.base.vcs_models`.

@mathjazz Can you take a look at this again?
@mythmon If you could give some feedback on this too that'd be great. :D

As a side note, I'm willing to land and test on some projects without automated tests to see if it solves the issues we're currently having, but my next PR will be a followup with automated tests.

Osmose commented:
Oh snap I totally forgot to add the logging you asked for

Osmose commented:
There, updated based on the etherpad.

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,363 @@
+        Generate a key for the given entity that is unique within the
+        project.
+        """
+        return ':'.join([entity.resource.path, entity.string])
```
This will still not work for formats that use "key" as a unique string identifier in a resource file (as opposed to "string").

mythmon commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,363 @@
+                    user,
+                    db_project.repository_url
+                )
+            except Exception as err:
```
This gives me flashbacks to overly zealous exception handlers that disguised typos in plugins and made things a lot harder to debug. Is it possible to narrow the caught types of exceptions?

mythmon commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,363 @@
+        if self.verbosity >= 1:
+            self.log(msg, *args, **kwargs)
+
+    def debug(self, msg, *args, **kwargs):
```
These are logging methods probably interesting enough to move into something like a `PontoonBaseCommand` class at some point.

mathjazz commented on pontoon/administration/management/commands/sync_projects.py:
```diff
@@ -0,0 +1,363 @@
+    def commit_changes(self, db_project):
+        """Commit the changes we've made back to the VCS."""
+        # TODO: Set translation author properly.
+        user = User.objects.filter(is_superuser=True).first()
```
I'd clal this a blocker, because it's very important to give credit to localizers also in VCS commits.

Osmose replied:
What makes me pause about this is that the old logic is to pick the most recent translator. We're not actually giving credit to every translator via VCS commits, just the one lucky enough to get approved last before they get synced.

Perhaps we can list all the authors in the commit message?

mathjazz replied:
Giving credit to the recent most translator is good enough, because there aren't many occasions when two localizers submit translations for the same project since last sync.

If there's more than one translator, listing the rest of them in the commit message would be better.

And the best we can do is probably this:
https://bugzilla.mozilla.org/show_bug.cgi?id=1179213

mythmon commented on pontoon/base/models.py:
```diff
@@ -255,6 +368,23 @@ def add_translation(self, string, locale, user):
+
+def extra_default():
+    """Default value for the Translation.extra field."""
+    return {}
```
This seems a bit silly. Why not just `extra = JSONFiled(default=lambda: {})`? Putting it out in a separate function makes me wonder what is so fancy about this function while I'm reading the model definition.

Osmose replied:
See https://docs.djangoproject.com/en/1.8/ref/models/fields/#default. Lambdas can't be serialized for migrations, nor can methods that aren't defined in the top level of the module. Thus, I use a top-level function that can be serialized.

mythmon replied:
Ah. That makes sense. I didn't know that.

mythmon commented on pontoon/base/vcs_models.py:
```diff
@@ -0,0 +1,127 @@
+from itertools import chain
+
+
+class VCSProject(object):
```
Is it normal in Pontoon to have models that aren't Django models?

Osmose replied:
Not until now!

I mean, they are undoubtly modeling the data that we work with, and they are a nice counterpoint to the DB-based models. But if your point is that calling them models is potentially misleading, I see that. I'm not sure if there's a better name for them than models, though.

Osmose commented:
Updated based on feedback.

@mathjazz How's it look now?

Osmose commented:
Updated with commit author fix.

I also found a possible blocker: The l10n langfile parser doesn't preserve linebreaks properly; it assumes everything has two line breaks between it. I'm working on a fix.

Osmose commented:
Updated, whitespace is now preserved in lang files.

Osmose commented:
@mathjazz r?

mathjazz commented:
r+

As I already said, this is great and can't wait for it!
